Security

tlatmsrud opened the pull request.

tlatmsrud committed:
first commit

강의 코드 실습

tlatmsrud committed:
강의 실습

tlatmsrud committed:
유저 정보 수정 권한에 대한 테스트

tlatmsrud committed:
본인 여부 확인 테스트 코드 작성

tlatmsrud committed:
본인 여부 확인 메서드를 구현하라

tlatmsrud committed:
UserService 리팩토링

UpdateUser 메서드에 user.changePassword 메서드 추가

tlatmsrud committed:
UserServiceTest 코드 리팩토링

tlatmsrud committed:
AuthenticationServiceTest 코드 리팩토링

tlatmsrud committed:
ControllerErrorAdvice의 미사용 메서드 삭제

tlatmsrud committed:
application.yml의 jpa.hibernate.ddl-auto을 craete로 변경

update로 되어있어 테스트가 계속 실패하였음.

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -5,32 +5,44 @@
         return claims.get("userId", Long.class);
     }
+
+    public boolean isPrincipal(Long userId) {
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        return authentication.getPrincipal().equals(userId);
+    }
```
전역 객체에서 가져오는 것 같네요. 전역 객체에 의존하니 테스트도 작성하기 어렵겠네요

tlatmsrud replied:
넵 테스트를 꾸역꾸역 작성하긴했는데 시간이 오래걸리긴 했었습니다. ㅠㅠ 말씀하신대로 전역객체에 의존하는 건 좋은 방법이 아닌 것 같아 @PreAuthorize("isAuthenticated() and getPrincipal() == #id") 구문을 사용하여 처리하였습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityConfig.java:
```diff
@@ -0,0 +1,45 @@
+
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+        http
+                .csrf().disable()
```
csrf란 무엇이고 여기서는 왜 disable했는지 알아봅시다

tlatmsrud replied:
CSRF 공격을 막기 위한 방법으로, 페이지 진입(GET) 시 서버로부터 CSRF 토큰을 발급받고 해당 토큰 값을 헤더나 바디에 셋팅하여 요청하므로써 '나는 대상 서버로부터 CSRF 토큰을 발급받은 검증된 사이트다' 라는 걸 알려주는 걸로 알고있습니다. 스프링 시큐리티를 적용하면 CSRF 토큰 방식이 기본으로 활성화되며 자동으로 CSRF Filter가 추가됩니다. 이 때 요청 바디나 헤더에 CSRF 토큰이 없다면 POST, FETCH, UPDATE, DELETE 따위의 API 호출 시 Access Denied 예외를 떨구는 관계로 임시적으로 disable 처리하였습니다

tlatmsrud committed:
코드 리팩토링

AuthenticationService의 isPrincipal 메서드는 전역객체에 의존하므로 이를
제거하고 Controller의 @PreAuthorize를 사용하여 getPrincipal() 과 요청 id
값을 비교하는 방법으로 변경하였음.

tlatmsrud commented:
멘토님, 혹시 현 코드에서 스프링 시큐리티 관련해서 보충하거나 추가할만한 기능을 제시해주실 수 있으신가요??

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityConfig.java:
```diff
@@ -0,0 +1,45 @@
+                .csrf().disable()
+                .addFilter(authenticationFilter)
+                .addFilterBefore(authenticationErrorFilter,JwtAuthenticationFilter.class)
+                .sessionManagement()
+                    .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
```
여기서 `STATELESS`를 사용하고 있군요. 이것의 의미는 무엇인가요?

tlatmsrud replied:
스프링 시큐리티를 통해 인증 완료 시 SecurityContext 에 인증 정보가 들어있는 Authenticaiton 객체를 저장합니다. 그리고 이 SecurityContext를 세션에 저장하는데, SecurityContextPersistenceFilter에서 세션 내 SecurityContext가 존재하면 인증 관련 필터를 건너 뛰게됩니다. 즉, 사용자의 인증을 Session으로 체크하는 것이 기본 설정입니다.
 하지만 지금은 STATELESS 방식 중 하나인 JWT로 인증을 하고 있기에 Session 체크가 필요 없으며, 이에따라 세션 정책을 STATELESS로 하여 SecurityContext를 세션에 저장하지 않도록 한다는 의미로 알고있습니다.

hannut91 commented:
지금은 Access Token만 가지고 있는데, Refresh token도 구현해 볼 수 있을 것 같고, Role도 지금은 사용자 밖에 없는데, 관리자나 다른 Role생기는 경우도 적용해 볼 수 있겠습니다

tlatmsrud committed:
코드 리팩토링 - 리프레시 토큰 적용

1. 로그인 시 accessToken과 refreshToken에 대한 만료기간 설정 및 생성 후
   리턴
2. 계정 생성 시 기본 role을 "ROLE_USER" 권한으로 생성

tlatmsrud committed:
계정 삭제 시 ROLE_ADMIN 권한 추가

tlatmsrud committed:
refreshToken을 통해 accessToken을 재발급받는 테스트코드 구현

1. 클라이언트에서 accessToken 만료 확인을  위한 예외 로직 추가
2. JWT 디코딩 시 ExpiredJwtException 발생 시 예외처리
3. SessionController의 accessToken 재발급 테스트코드 생성

tlatmsrud committed:
리프레시 토큰을 통한 액세스 토큰 재발급 API 개발

tlatmsrud commented:
> 지금은 Access Token만 가지고 있는데, Refresh token도 구현해 볼 수 있을 것 같고, Role도 지금은 사용자 밖에 없는데, 관리자나 다른 Role생기는 경우도 적용해 볼 수 있겠습니다

1. 로그인 시 AccessToken과 RefreshToken을 만료 시간만 다르게하여 발급하였으며, 이 시점에 DB에 해당 유저에 대한 RefreshToken을 갱신시켰습니다.

2. SessionController에 RefreshToken을 통해 AccessToken을 갱신하여 응답해주는 reissue  메서드를 생성하였습니다.

3. ADMIN 권한을 가진 JWT 토큰에 대해서만 유저 삭제가 되도록 처리해보았습니다.

4. 클라이언트가 본인이 요청한 AcccessToken 만료 여부를 확인할 수 있도록 jwtUtil.decode에 작업 하였으며, 클라이언트에게 401에러와 토큰이 만료되었다는 메시지를 AuthenticationErrorFilter에 추가하였으나,  특이사항이 발생하여 질문드립니다.
 
>> 테스트코드에서 임시로 만료된 토큰을 만든 후 인증 정보가 필요한 post products/ API를 호출하였으나, 401 에러만 발생하고 
  메시지는 받을 수 없었습니다. 디버깅 결과 @PreAuthorize가 있을 경우 Filter 호출 전에 인증 정보를 체크하는 것으로 확인되었으며, 이때 만료된 토큰이 들어왔으므로 401 에러를 내려주고 있었습니다. 필터를 거쳤다면 토큰이 만료되었다는 메시지를 클라한테 내려줬을텐데 말이죠.. 궁금한 점은 무상태 프로토콜인 JWT의 인증, 인가 정보는 JwtAuthenticationFilter에서 HTTP 요청마다 생성하고 Authentication 객체를 만들어 SecurityContext에 넣어주고있는데, Filter를 거치기 전에 어떻게 이 토큰의 만료 여부를 확인했는지 궁금합니다. 

>> 그리고 JWT 리프레시 토큰은 쿠키에 넣어놓고 httpOnly, Secure 설정을하고, 액세스 토큰은 클라이언트에서 로컬변수로 가지고 있는게 그나마 안전하다들 하던데 멘토님은 어떤 방식으로 리프레시 토큰과 액세스 토큰을 관리하시나요?

tlatmsrud committed:
Token 유효기간 변경

hannut91 commented:
쿠키에 넣고, httpOnly로 해야 XSS공격에도 세션이 탈취되지 않도록 하는게 보편적으로 안전하다고 하죠. 그렇지만 애초에 XSS공격을 막는게 더 중요하다고 생각해서 저는 클라이언트에서 저장합니다.

또 어떤 곳에서는 Refresh Token만 서버에서 데이터 베이스에 관리하기도 합니다.

또 다른 곳에서는 보안에 아주 중요하지 않은 부분들은 JWT로 인증하다가 결제나 다른 중요한 것이 있을 때는 서버의 세션을 이용해서 처리하는 경우도 있습니다. 우리가 서비스를 잘 사용하고 있다가 특정 기능을 사용하려고 하다 보면 갑자기 다시 인증을 확인하라는 요청이 뜬 것을 본적이 있으실거예요.

tlatmsrud committed:
코드 리팩토링

tlatmsrud commented:
테스트코드에서 임시로 만료된 토큰을 만든 후 인증 정보가 필요한 post products/ API를 호출하였으나, 401 에러만 발생하고
메시지는 받을 수 없었습니다. 디버깅 결과 @PreAuthorize가 있을 경우 Filter 호출 전에 인증 정보를 체크하는 것으로 확인되었으며, 이때 만료된 토큰이 들어왔으므로 401 에러를 내려주고 있었습니다. 필터를 거쳤다면 토큰이 만료되었다는 메시지를 클라한테 내려줬을텐데 말이죠.. 궁금한 점은 무상태 프로토콜인 JWT의 인증, 인가 정보는 JwtAuthenticationFilter에서 HTTP 요청마다 생성하고 Authentication 객체를 만들어 SecurityContext에 넣어주고있는데, Filter를 거치기 전에 어떻게 이 토큰의 만료 여부를 확인했는지 궁금합니다.

>> 오늘 인텔리제이 재시작하고 다시 해보니 정상적으로 돌아갑니다. ㅠㅠ 감사합니다!
